experimental/Schemes/elliptic_surface.jl

```diff
@@ -90,15 +102,84 @@ Elliptic surface
 with generic fiber
   -x^3 + y^2 - t^7 + 2*t^6 - t^5
 
-julia> Base.show(stdout, X3)
-Elliptic surface with generic fiber -x^3 + y^2 - t^7 + 2*t^6 - t^5
 ```
 """
-function elliptic_surface(generic_fiber::EllipticCurve{BaseField}, euler_characteristic::Int,
-                          mwl_basis::Vector{<:EllipticCurvePoint}=EllipticCurvePoint[]) where {
+function elliptic_surface(generic_fiber::EllipticCurve{BaseField},
+                          euler_characteristic::Int,
+                          mwl_gens::Vector{<:EllipticCurvePoint}=EllipticCurvePoint[];
+                          is_basis::Bool=true) where {
                           BaseField <: FracFieldElem{<:PolyRingElem{<:FieldElem}}}
-  @req all(parent(i)==generic_fiber for i in mwl_basis) "not a vector of points on $(generic_fiber)"
-  S = EllipticSurface(generic_fiber, euler_characteristic, mwl_basis)
+  @req all(parent(i)==generic_fiber for i in mwl_gens) "not a vector of points on $(generic_fiber)"
+  S = EllipticSurface(generic_fiber, euler_characteristic, mwl_gens)
+  if is_basis
+    return S
+  end
+  update_mwl_basis!(S, mwl_gens)
+  return S
+end
+
+@doc raw"""
+    update_mwl_basis!(S::EllipticSurface, mwl_gens::Vector{<:EllipticCurvePoint})
+
+Compute a reduced basis of the sublattice of the Mordell-Weil lattice spanned
+by `mwl_gens` and set these as the new generators of the Mordell-Weil lattice of
+`S`.
+"""
+function update_mwl_basis!(S::EllipticSurface, mwl_gens::Vector{<:EllipticCurvePoint})
+  mwl, mwl_basis = _compute_mwl_basis(S, mwl_gens)
+  set_mordell_weil_basis!(S, mwl_basis)
+end
+
+@doc raw"""
+    algebraic_lattice_primitive_closure(S::EllipticSurface, p) -> Vector{<:EllipticCurvePoint}
+
+Return sections ``P_1,\dots P_n`` of the generic fiber, such that together with
+the generators of the algebraic lattice ``A``, they generate
+``(1/p A \cap N)`` where ``N`` is the numerical lattice of ``S``.
+
+This proceeds by computing division points in the Mordell-Weil group
+and using information coming from the discriminant group of the algebraic lattice
+to do so.
+"""
+function algebraic_lattice_primitive_closure(S::EllipticSurface, p)
```

Comment:
Don't you want to specify some type for p? You want to divide by it later on, so you might want to protect yourself from accidental misleading input.